### ISSUE TITLE
sysext: does not support submounts below /usr/. Should be documented, warned, or alternatively made work

### ISSUE THREAD

**Issue by hsitter** [systemd/systemd#31922]
### systemd version the issue has been seen with

255.4-2-arch

### Used distribution

Arch

### Linux kernel version used

6.8.1-arch1-1

### CPU architectures issue was seen on

x86_64

### Component

systemd-sysext

### Expected behaviour you didn't see

Having a pre-existing mount in /usr should be preserved after sysext merge.

### Unexpected behaviour you saw

Using sysext on a /usr that already contains mounts loses those extra mounts. The files from the real /usr leak through.

### Steps to reproduce the problem

```
mount -t tmpfs tmpfs /usr/lib/locale
touch /usr/lib/locale/foo
systemd-sysext refresh # of course needs some extensions installed
ls /usr/lib/locale # no foo in sight
```

### Additional program output to the terminal or log subsystem illustrating the issue

```sh
My specific setup has a read-only /usr that needs a writable /usr/lib/locale mounted into for locale generation support. When that /usr gets sysextended the locales get lost and my glibc becomes sad.

Essentially I have


# read-only subvolume of the OS
/dev/nvme1n1p2 on / type btrfs (rw,nodev,relatime,ssd,discard=async,space_cache,subvolid=6500,subvol=/@os_202403232030)
# writable mount on top
/dev/nvme1n1p2 on /usr/lib/locale type btrfs (rw,relatime,ssd,discard=async,space_cache,subvolid=6496,subvol=/@locale)
# extensions
sysext on /usr type overlay (ro,nodev,relatime,lowerdir=/run/systemd/sysext/meta/usr:/run/systemd/sysext/extensions/zxing/usr:/run/systemd/sysext/extensions/yay/usr:/run/systemd/sysext/extensions/workspace/usr:/run/systemd/sysext/extensions/wezterm/usr:/run/systemd/sysext/extensions/spectacle/usr:/run/systemd/sysext/extensions/plasma-pa/usr:/run/systemd/sysext/extensions/kde/usr:/run/systemd/sysext/extensions/hstr/usr:/run/systemd/sysext/extensions/core/usr:/usr,index=off)
```
```

**Comment by bluca** [comment]
This is not supported, the base /usr/ is supposed to originate from a single filesystem, either the rootfs or a /usr/ specific one.

**Comment by hsitter** [comment]
That's fair. I've tried attaching a mount unit to usr.mount instead so the locale always gets mounted on top, but that for some reason doesn't work. Is that a bug?

```
# cat /etc/systemd/system/usr-lib-locale.mount 
[Unit]
Description=Locale Directory Mount
After=usr.mount

[Mount]
What=/dev/gpt-auto-root
Where=/usr/lib/locale
Options=rw,subvol=@locale
LazyUnmount=yes
ReadWriteOnly=yes

[Install]
WantedBy=usr.mount


# readlink /etc/systemd/system/usr.mount.wants/usr-lib-locale.mount 
/etc/systemd/system/usr-lib-locale.mount


# systemctl daemon-reload


# systemd-sysext refresh
Using extensions 'core', 'hstr', 'kde', 'plasma-pa', 'spectacle', 'wezterm', 'workspace', 'yay', 'zxing'.
Merged extensions into '/usr'.


# systemctl is-active usr.mount
active


# systemctl is-active usr-lib-locale.mount
inactive
```

I see the unit changing state in the debug log

```
Mar 24 18:19:46 ajax systemd[1]: usr.mount: Changed dead -> mounted
```

but somehow it doesn't queue wants nor requires relationships.

**Comment by poettering** [comment]
I guess we should extend the man page and say clearly that submounts are not supported, currently. maybe even fail with a clear error when this is detected.

I guess we could make this work if we wanted, i.e. whenever we put together a new overlayfs bind mound all submounts over first, before moving it into place. I am a bit too lazy to implement that however, and I don't really see the usecase.

**Comment by hsitter** [comment]
I mean, ignoring the lack of documentation, the status quo is OK.

Iff usr.mount changes actually propagated through the unit system that is. Then one can re-establish mounts on top of the newly overlayed usr but with the propagation also not working it's a bit unfortunate.

Indeed I've hit a similar problem just now again where I want to reload dbus-broker upon `systemd-sysext refresh` but I can't because neither usr.mount nor systemd-sysext.service seem to trigger any relation.

**Comment by poettering** [comment]
Fixed by ddbddebe0cc5edcf3d0fbc82ba5850e45925e8ae
